src/ui/public/agg_types/__tests__/buckets/terms.js

```diff
@@ -0,0 +1,156 @@
+      ngMock.module('kibana');
+      ngMock.inject(function (Private, $controller, _$rootScope_) {
+        const terms = Private(AggTypesIndexProvider).byName.terms;
+        const orderAggController = terms.params.byName.orderAgg.controller;
+
+        $rootScope = _$rootScope_;
+        $rootScope.agg = {
+          id: 'test',
+          params: {},
+          type: terms,
+          vis: {
+            aggs: []
+          }
+        };
+        $rootScope.responseValueAggs = responseValueAggs;
+        $controller(orderAggController, { $scope: $rootScope });
+        $rootScope.$digest();
+      });
+    }
+
+    it('defaults to the first metric agg', function () {
+      init({
+        responseValueAggs: [
+          {
+            id: 'agg1',
+            type: {
+              name: 'count'
+            }
+          },
+          {
+            id: 'agg2',
+            type: {
+              name: 'count'
+            }
+          }
+        ]
+      });
+      expect($rootScope.agg.params.orderBy).to.be('agg1');
+    });
+
+    it('defaults to the first metric agg that is compatible with the terms bucket', function () {
+      init({
+        responseValueAggs: [
+          {
+            id: 'agg1',
+            type: {
+              name: 'top_hits'
+            }
+          },
+          {
+            id: 'agg2',
+            type: {
+              name: 'percentiles'
+            }
+          },
+          {
+            id: 'agg3',
+            type: {
+              name: 'median'
+            }
+          },
+          {
+            id: 'agg4',
+            type: {
+              name: 'std_dev'
+            }
+          },
+          {
+            id: 'agg5',
+            type: {
+              name: 'count'
+            }
+          }
+        ]
+      });
+      expect($rootScope.agg.params.orderBy).to.be('agg5');
+    });
+
+    it('defaults to the custom metric if no agg is compatible', function () {
+      init({
+        responseValueAggs: [
+          {
+            id: 'agg1',
+            type: {
+              name: 'top_hits'
+            }
+          }
+        ]
+      });
+      expect($rootScope.agg.params.orderBy).to.be('custom');
+    });
+
+    it('selects "custom metric" if there are no metric aggs', function () {
+      init({});
+      expect($rootScope.agg.params.orderBy).to.be('custom');
+    });
+
+    it('is emptied if the selected metric becomes incompatible', function () {
+      init({
+        responseValueAggs: [
+          {
+            id: 'agg1',
+            type: {
+              name: 'count'
+            }
+          }
+        ]
+      });
+      expect($rootScope.agg.params.orderBy).to.be('agg1');
+      $rootScope.responseValueAggs = [
+        {
+          id: 'agg1',
+          type: {
+            name: 'top_hits'
+          }
+        }
+      ];
+      $rootScope.$digest();
+      expect($rootScope.agg.params.orderBy).to.be(null);
```

Comment:
I think I understand your thinking here, if someone selected a metric and then it becomes incompatible, you'd like to force them to make a new selection, is that correct? The downside to this is that every time someone changes the default `count` metric to top, it clears out the order and they have to manually select it, whereas I think defaulting to sort by `Term` would be reasonable. 

Instead of setting it to null, why don't we apply the same default setting logic that would get applied when the agg is first loaded? In other words, if the order metric is removed or become incompatible do the following:

1. Default to the first compatible metrics
2. If no compatible metric exists, default to Custom or Term (I would recommend Term).